srai/embedders/geovex/embedder.py

```diff
@@ -0,0 +1,244 @@
+"""
+GeoVex Embedder.
+
+This module contains embedder from GeoVex paper[1].
+
+References:
+    [1] https://openreview.net/forum?id=7bvWopYY1H
+"""
+from typing import Any, Dict, List, Optional, Tuple, TypeVar
+
+import geopandas as gpd
+import numpy as np
+import pandas as pd
+
+from srai.constants import REGIONS_INDEX
+from srai.embedders import CountEmbedder
+from srai.embedders.geovex.dataset import HexagonalDataset
+from srai.embedders.geovex.model import GeoVexModel
+from srai.exceptions import ModelNotFitException
+from srai.neighbourhoods import H3Neighbourhood
+from srai.utils._optional import import_optional_dependencies
+
+T = TypeVar("T")
+
+try:  # pragma: no cover
+    from torch.utils.data import DataLoader
+
+except ImportError:
+    from srai.utils._pytorch_stubs import DataLoader
+
+
+class GeoVexEmbedder(CountEmbedder):
+    """Hex2Vec Embedder."""
+
+    def __init__(
+        self,
+        target_features: List[str],
+        neighbourhood: H3Neighbourhood,
+        batch_size: Optional[int] = 32,
+        neighbourhood_radius: int = 4,
+        convolutional_layers: int = 2,
+        embedding_size: int = 32,
+        dataset: Optional[HexagonalDataset[T]] = None,
+    ) -> None:
+        """
+        Initialize GeoVex Embedder.
+
+        Args:
+            target_features (List[str]): The features that are to be used in the embedding.
+                Should be in "flat" format, i.e. "<super tag>_<sub_tag>".
+            neighbourhood (H3Neighbourhood): The neighbourhood to use.
+                Should be intialized with the same regions.
+            batch_size (int, optional): Batch size. Defaults to 32.
+            convolutional_layers (int, optional): Number of convolutional layers. Defaults to 2.
+            neighbourhood_radius (int, optional): Radius of the neighbourhood. Defaults to 4.
+            embedding_size (int, optional): Size of the embedding. Defaults to 32.
+            dataset (Optional[HexagonalDataset], optional): Dataset to use. Defaults to None.
+        """
+        import_optional_dependencies(
+            dependency_group="torch", modules=["torch", "pytorch_lightning"]
+        )
+
+        super().__init__(
+            expected_output_features=target_features,
+        )
+
+        self._model: Optional[GeoVexModel] = None
+        self._is_fitted = False
+        self._r = neighbourhood_radius
+        self._embedding_size = embedding_size
+        self._convolutional_layers = convolutional_layers
+
+        self._neighbourhood = neighbourhood
+        self._batch_size = batch_size
+        self._neighbourhood = neighbourhood
+        self._dataset = dataset
+
+        # save invalid h3s for later
+        self._invalid_h3s: List[str] = []
+
+    def transform(
+        self,
+        regions_gdf: gpd.GeoDataFrame,
+        features_gdf: gpd.GeoDataFrame,
+        joint_gdf: gpd.GeoDataFrame,
+    ) -> pd.DataFrame:
```

Comment:
Good observation. Will fix. 

FWIW, as an outsider who has only used this library to implement the GeoVex example, the need for 3 dataframes + a neighborhood object to describe a region feels a bit excessive? Seems like there is potentially a more concise way to represent an area?

Reply:
Well, you're right on that. We have actually already drafted a proposal to replace it with a single object, but it is WIP. Maybe @simonusher can tell a bit more about how we plan to approach this

Reply:
Awesome! Also mypy through an error when I added the neighborhood as an argument to this function. Instead I just added the re-creation of the neighborhood object inside of the function.

Reply:
Hey, I'm very sorry for not responding to this earlier. I agree with what you said -> the API is a bit cumbersome and we will be working to improve it. We overlooked some use cases when designing it, then added them afterwards. At first, we also thought this would be more of an internal API and that we would write an easier one on top of it (which will probably happen anyway).
The way we approach this will probably also be based on the results we have with using dask for out-of-core processing and trying out duckdb.
What I've been playing with was just having a container object like [HF's BatchEncoding](https://huggingface.co/docs/transformers/v4.34.0/en/main_classes/tokenizer#transformers.BatchEncoding) but I'm not sure if we'll go this exact way. Anyway, this requires some more design work. Feel free to share any ideas if you'd like :)

Reply:
>  We overlooked some use cases when designing it, then added them afterwards. At first, we also thought this would be more of an internal API and that we would write an easier one on top of it (which will probably happen anyway).

Totally understand that! 

Re dask, when we implemented your Hex2Vec model, we were training on city metro-regions, and had to switch to dask out-of-core for the spatial join of the larger regions and had success. 

A database like DuckDB could be the best solution, where the relationship between regions, features, joint, and ultimately embeddings happens behind the scenes so to say. Then again ig there is a fine line to walk between accessibility, ease-of-use, and performance